### ISSUE TITLE
Unnecessary huge build artifact directory despite there is only one distinct CodeUri

### ISSUE THREAD

**Issue by aahung** [aws/aws-sam-cli#3820]
### Describe your idea/feature/enhancement

I have a SAM template with 100+ lambda functions, all share the same `CodeUri`. Similar to #3301 which illustrates the duplicate artifact uploads, `sam build` is very slow and it generates 10+GB of build artifacts, while each function only has 90MB build artifact. 

```
~/P/d/s/.aws-sam ❯❯❯ du -csh *
1.4M	auto-dependency-layer
 10G	build
8.0K	build.toml
316M	cache
303M	deps
 11G	total
```

The root cause is during `sam build`, SAM CLI will copy the build artifact of functions sharing the same code hash to `.aws-sam/build/<function-logical-id>`. If there are 100+ functions, SAM CLI will do it for 100+ times, which waste a lot of space and cause excessive SSD wear.

Related code:

https://github.com/aws/aws-sam-cli/blob/23beec68bf826ab646344be2cd858b1b4cf4ac7e/samcli/lib/build/build_strategy.py#L172-L180

https://github.com/aws/aws-sam-cli/blob/23beec68bf826ab646344be2cd858b1b4cf4ac7e/samcli/lib/build/build_strategy.py#L281-L286

### Proposal

- Get rid of the build directory naming convention `.aws-sam/build/<function-logical-id>` and favor `.aws-sam/build/<function-code-hash>`, so functions sharing the same code hash only need one build directory.
- in `sam deploy` or `sam local` which depends on `.aws-sam/build/<function-logical-id>`, use the new `.aws-sam/build/<function-code-hash>`. The corresponding mapping is available in `build.toml` (`source_hash` -> `functions` under `function_build_definitions.*`)

Things to consider:
1. Will this require any updates to the [SAM Spec](https://github.com/awslabs/serverless-application-model)

**Comment by aahung** [comment]
# Quick Workaround using symlink instead of `osutils.copytree`

Wanna share a quick workaround, only tested on macOS. It might not handle some corner cases.

at line 178, change to

```py
LOG.debug("Symlink artifacts from %s to %s", single_build_dir, artifacts_dir)
pathlib.Path(artifacts_dir).parent.mkdir(exist_ok=True)
os.symlink(single_build_dir, artifacts_dir)
```

at line 284, similarly: 
```py
LOG.debug("Symlink artifacts from %s to %s", cache_function_dir, artifacts_dir)
pathlib.Path(artifacts_dir).parent.mkdir(exist_ok=True)
os.symlink(cache_function_dir, artifacts_dir)
```

## Before
`du -csh *` in `.aws-sam`: 
```
1.4M	auto-dependency-layer
 10G	build
8.0K	build.toml
316M	cache
303M	deps
 11G	total
```
`time sam build --use-container --cached --parallel --skip-pull-image --debug`: 
```
12.51s user 69.35s system 92% cpu 1:28.09 total
```
## After

`du -csh *` in `.aws-sam`: 
```
1.4M	auto-dependency-layer
172K	build
8.0K	build.toml
316M	cache
303M	deps
620M	total
```
`time sam build --use-container --cached --parallel --skip-pull-image --debug`: 
```
2.68s user 0.48s system 69% cpu 4.549 total
```

**Comment by moelasmar** [comment]
Thanks @aahung for raising this Feature request.

**Comment by super132** [comment]
This feature was implemented as a beta feature and released in `1.54.0`. I'm resolving this issue. Thanks.
